File: app/src/main/java/de/christinecoenen/code/zapp/app/mediathek/controller/DownloadReceiver.java
package de.christinecoenen.code.zapp.app.mediathek.controller;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import de.christinecoenen.code.zapp.app.ZappApplication;
import de.christinecoenen.code.zapp.models.shows.PersistedMediathekShow;
import de.christinecoenen.code.zapp.repositories.MediathekRepository;
import de.christinecoenen.code.zapp.app.mediathek.ui.detail.MediathekDetailActivity;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import timber.log.Timber;


public class DownloadReceiver extends BroadcastReceiver {

	private static final String ACTION_NOTIFICATION_CLICKED = "de.christinecoenen.code.zapp.NOTIFICATION_CLICKED";
	private static final String EXTRA_DOWNLOAD_ID = "EXTRA_DOWNLOAD_ID";

	public static Intent getNotificationClickedIntent(Context context, int downloadId) {
		Intent intent = new Intent(context, DownloadReceiver.class);
		intent.setAction(ACTION_NOTIFICATION_CLICKED);
		intent.putExtra(EXTRA_DOWNLOAD_ID, downloadId);
		return intent;
	}

	@Override
	public void onReceive(Context context, Intent intent) {
		if (context == null || intent == null) {
			return;
		}

		String action = intent.getAction();

		if (ACTION_NOTIFICATION_CLICKED.equals(action)) {

			ZappApplication application = (ZappApplication) context.getApplicationContext();
			MediathekRepository mediathekRepository = application.getMediathekRepository();
			int downloadId = intent.getIntExtra(EXTRA_DOWNLOAD_ID, 0);

			Disposable loadShowDisposable = mediathekRepository
				.getPersistedShowByDownloadId(downloadId)
				.firstElement()
				.observeOn(AndroidSchedulers.mainThread())
				.subscribe(persistedMediathekShow -> onShowLoaded(context, persistedMediathekShow), Timber::e);
		}
	}

	private void onShowLoaded(Context context, PersistedMediathekShow persistedMediathekShow) {
		// launch MediathekDetailActivity
		Intent detailIntent = MediathekDetailActivity.getStartIntent(context, persistedMediathekShow.getMediathekShow());
		detailIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
		detailIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
		context.startActivity(detailIntent);
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/mediathek/controller/DownloadReceiver.kt
package de.christinecoenen.code.zapp.app.mediathek.controller

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.navigation.NavDeepLinkBuilder
import de.christinecoenen.code.zapp.R
import de.christinecoenen.code.zapp.models.shows.PersistedMediathekShow
import de.christinecoenen.code.zapp.repositories.MediathekRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

class DownloadReceiver : BroadcastReceiver(), KoinComponent {

	companion object {

		private const val ACTION_NOTIFICATION_CLICKED =
			"de.christinecoenen.code.zapp.NOTIFICATION_CLICKED"
		private const val EXTRA_DOWNLOAD_ID = "EXTRA_DOWNLOAD_ID"

		fun getNotificationClickedIntent(context: Context?, downloadId: Int): Intent {
			return Intent(context, DownloadReceiver::class.java).apply {
				action = ACTION_NOTIFICATION_CLICKED
				putExtra(EXTRA_DOWNLOAD_ID, downloadId)
			}
		}

	}

	private val mediathekRepository: MediathekRepository by inject()
	private val coroutineScope: CoroutineScope by inject()

	@SuppressLint("CheckResult")
	override fun onReceive(context: Context, intent: Intent) {
		if (ACTION_NOTIFICATION_CLICKED != intent.action) {
			return
		}

		val downloadId = intent.getIntExtra(EXTRA_DOWNLOAD_ID, 0)

		coroutineScope.launch {
			val persistedShow = mediathekRepository
				.getPersistedShowByDownloadId(downloadId)
				.first()

			onShowLoaded(context, persistedShow)
		}
	}

	private fun onShowLoaded(context: Context, persistedMediathekShow: PersistedMediathekShow) {
		NavDeepLinkBuilder(context)
			.setGraph(R.navigation.nav_graph)
			.setDestination(R.id.mediathekDetailFragment)
			.setArguments(Bundle().apply {
				putSerializable("mediathek_show", persistedMediathekShow.mediathekShow)
			})
			.createPendingIntent()
			.send()
	}
}
